### PR TITLE
Add Pattern Match and Rewrite Library For Extensions

### PR DESCRIPTION
**Pull request by silverstar194**
Signed-off-by: silverstar195 <seanmaloney@google.com>

Commit Message:
Add Pattern Match and Rewrite Library For Extensions Additional Description:

Additional Description:
Added extension library for implementation of url pattern matching and rewrite.

This library will be called by both extensions as needed.

Full set of changes to be merged is mostly complete and tested. It does still need cleaning up. I'll be splitting PRs off of it here:
https://github.com/envoyproxy/envoy/pull/22207

Risk Level:
No changes to existing functionality within this PR.

Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/path/pattern_template_lib/proto/pattern_template_rewrite_segments.proto [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.pattern_template";
+option java_outer_classname = "PatternTemplateRewrite";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/pattern_template/pattern_template";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Pattern Template Rewrite]
```
Comment (phlax) [inline source/extensions/path/pattern_template_lib/proto/pattern_template_rewrite_segments.proto:13]: perhaps im missing something - but to be used by protodoc this needs to be in the api

Comment (silverstar194) [reply source/extensions/path/pattern_template_lib/proto/pattern_template_rewrite_segments.proto:13]: It is internal to the extension. No need to have user facing docs

Comment (phlax) [reply source/extensions/path/pattern_template_lib/proto/pattern_template_rewrite_segments.proto:13]: in which case we dont need the protodoc annotations

Comment (silverstar194) [reply source/extensions/path/pattern_template_lib/proto/pattern_template_rewrite_segments.proto:13]: Ok will remove

File: CODEOWNERS [hunk 2]
```diff
@@ -257,6 +257,10 @@ extensions/filters/http/oauth2 @derekargueta @snowp
 # Thrift
 /*/extensions/filters/network/thrift_proxy @zuercher @JuniorHsu
 
+# URL Pattern Match and Rewrite
```
Comment (silverstar194) [inline CODEOWNERS:260]: Who is the best owner for this?

Comment (alyssawilk) [reply CODEOWNERS:260]: as said on https://github.com/envoyproxy/envoy/pull/22459 you can put me in for now but we should see if @krajshiva can find someone from EPT to co-own

**Comment by alyssawilk** [comment]
/wait on CI.  
I think you're going to want to run format checks locally before pushing upstream - it'll save you a bunch of cycles

**Comment by silverstar194** [comment]
> /wait on CI. I think you're going to want to run format checks locally before pushing upstream - it'll save you a bunch of cycles

All presubmits are passing except. Tracking 2nd owner now
ERROR: Extensions require at least 2 owners in CODEOWNERS:
    /*/extensions/path/pattern_template_lib @alyssawilk

**Review by stevenzzzz — COMMENTED** [review]
Some nits. 

I think we want to see the config_impl.cc change PR first to understand better what this PR or  https://github.com/envoyproxy/envoy/pull/22459 are doing.

**Comment by alyssawilk** [comment]
Should we close this off in favor of the larger PR?
/wait

**Comment by silverstar194** [comment]
Closing out for full PR: https://github.com/envoyproxy/envoy/pull/22207

**Pull request closed without merging**